libdnf/advisory/advisory_query.cpp

```diff
@@ -0,0 +1,413 @@
+    }
+
+    // Apply filter results to query
+    if (cmp_not) {
+        data_map -= filter_result;
+    } else {
+        data_map &= filter_result;
+    }
+
+    return *this;
+}
+
+std::vector<AdvisoryPackage> AdvisoryQuery::get_advisory_packages(
+    const libdnf::rpm::PackageSet & package_set, sack::QueryCmp cmp_type) {
+    std::vector<AdvisoryPackage> pkgs = get_sorted_advisory_packages();
+    std::vector<AdvisoryPackage> after_filter;
+
+    auto solv_sack = package_set.get_sack();
+    Pool * pool = solv_sack->p_impl->get_pool();
+
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ:
+            //TODO(amatej): faster EQ specific version (we can compare whole NEVRA)
+        case libdnf::sack::QueryCmp::GT:
+        case libdnf::sack::QueryCmp::LT:
+        case libdnf::sack::QueryCmp::GTE:
+        case libdnf::sack::QueryCmp::LTE: {
+            for (libdnf::rpm::PackageSet::iterator package = package_set.begin(); package != package_set.end();
+                 package++) {
+                Solvable * solvable = libdnf::rpm::solv::get_solvable(pool, (*package).get_id());
+                auto low = std::lower_bound(pkgs.begin(), pkgs.end(), *package, AdvisoryPackage::comparePackageNA);
+                while (low != pkgs.end() && low->private_pkg.get()->get_name_id() == solvable->name &&
+                       low->private_pkg.get()->get_arch_id() == solvable->arch) {
+                    int libsolv_cmp =
+                        pool_evrcmp(pool, low->private_pkg.get()->get_evr_id(), solvable->evr, EVRCMP_COMPARE);
+                    if (((libsolv_cmp > 0) && ((cmp_type & sack::QueryCmp::GT) == sack::QueryCmp::GT)) ||
+                        ((libsolv_cmp < 0) && ((cmp_type & sack::QueryCmp::LT) == sack::QueryCmp::LT)) ||
+                        ((libsolv_cmp == 0) && ((cmp_type & sack::QueryCmp::EQ) == sack::QueryCmp::EQ))) {
+                        after_filter.push_back(*low);
+                    }
+                    ++low;
+                }
+            }
+        } break;
+        default:
+            throw NotSupportedCmpType("Unsupported CmpType");
+    }
+
+    //after_filter contains just advisoryPackages which comply to condition with package_set
+    return after_filter;
+}
+
+std::vector<Advisory> AdvisoryQuery::get_advisories() const {
+    std::vector<Advisory> out;
+    for (libdnf::advisory::AdvisoryId candidate_id : data_map) {
+        out.emplace_back(Advisory(*(advisory_sack->base), candidate_id));
+    }
+
+    return out;
+}
+
+std::vector<AdvisoryPackage> AdvisoryQuery::get_sorted_advisory_packages() const {
+    std::vector<AdvisoryPackage> out;
+    for (libdnf::rpm::PackageId candidate_id : data_map) {
+        Advisory advisory = Advisory(*(advisory_sack->base), AdvisoryId(candidate_id.id));
+        auto collections = advisory.get_collections();
+        for (const auto & collection : collections) {
+            collection.get_packages(out);
```

Comment:
It would probably make the most sense here? I have tried adding a bool parameter `only_applicable` (default to false) to the `get_sorted_advisory_packages` method.